### ISSUE TITLE
runtime, internal/poll: darwin: ensure that no thread is consumed, nor a syscall.Read if FD isn't yet ready for I/O

### ISSUE THREAD

**Issue by odeke-em** [golang/go#33953]
Coming here from https://github.com/golang/go/issues/32326.
That test was meant to ensure that an (*os.File).Read which is backed by internal/poll would ONLY consume a thread for a read syscall if I/O was ready.

Examining @MichaelTJones' crash dump in reveals in deed that when it crashed
```shell
$ grep 'os_test.go' stack | sort | uniq
/Users/mtj/go/src/os/os_test.go:2283 
/Users/mtj/go/src/os/os_test.go:2283 +0x370 
/Users/mtj/go/src/os/os_test.go:2286 +0x84 
/Users/mtj/go/src/os/os_test.go:2286 +0x84 fp=0xc00025afa8 sp=0xc00025af48 
/Users/mtj/go/src/os/os_test.go:2297 +0x3aa 
``` 
non of the code had gotten past blocking on a syscall.Read as per
https://github.com/golang/go/blob/e87fe0f1f521150641c8c64db6b919555c255c65/src/os/os_test.go#L2297

and also from his execution traces
![image](https://user-images.githubusercontent.com/4898263/63964928-fc41b280-ca4c-11e9-894b-1749dcfdb186.png)

Somehow the poller is tripping out and not blocking until the file descriptor is ready for I/O.

In that test, our high water mark for the max number of threads was 50 but for regular machines where GOMAXPROCS=2,4,8, the maximum value of threads if every call was doing a syscall.Read would be 10, 18, 34 if we follow
```go
        P := runtime.GOMAXPROCS(0)
        n := threads
        maxThreads := 0
        if n/P-1 >= P {
                maxThreads = 2 + 4*P
        } else {
                maxThreads = 1 + (7 * P / 4) + (n / P)
        }
``` 
but in his case, given GOMAXPROCS=32,36, it would be 61, 67.

Hence why his test failed but it went undetected for usually small values of GOMAXPROCS. Let's also examine the behaviors between versions of Go to see when things regressed.

**Comment by odeke-em** [comment]
I found the commit that introduced this problem, commit https://github.com/golang/go/commit/a3b01440fef3d2833909f6651455924a1c86d192#diff-38e21adab3460688e3ce0b71fa9422d0 aka CL https://go-review.googlesource.com/c/141639 which is a part of Go1.12.

This program should be able to run alright with GOMAXPROCS=8 and a max of 8 threads before that commit
```go
package main

import (
	"log"
	"os"
	"runtime/debug"
	"sync"
	"syscall"
	"time"
)

func main() {
	n := 100

	var rl, wl []*os.File
	for i := 0; i < n; i++ {
		prc, pwc, err := os.Pipe()
		if err != nil {
			// Cleanup.
			for j := 0; j < i; j++ {
				wl[i].Close()
				rl[i].Close()
			}
		}
		rl = append(rl, prc)
		wl = append(rl, pwc)
	}

	defer debug.SetMaxThreads(debug.SetMaxThreads(8))

	var wg sync.WaitGroup
	for i := 0; i < n; i++ {
		wg.Add(1)
		go func(i int) {
			bs := make([]byte, 4)
			wg.Done()
			prc := rl[i]
			if _, err := prc.Read(bs); err != nil {
				log.Fatalf("Failed to read %d: %v\n", i, err)
			}
		}(i)
	}

	wg.Wait()
	println("Waiting now")
	for {
		<-time.After(5 * time.Second)
		if true {
			return
		}
		proc, _ := os.FindProcess(os.Getpid())
		proc.Signal(syscall.SIGQUIT)
	}

	for i := 0; i < n; i++ {
		if _, err := wl[i].Write([]byte("Hello")); err != nil {
			log.Fatalf("Write #%d failed: %v", err)
		}
	}
}
```
Kindly pinging @randall77. 
For Go1.13, we've disabled the test on Darwin but after we've fixed the issue, we'll need to backport to Go1.12 and Go1.13 as well.

**Comment by odeke-em** [comment]
Some great news, @minux's CL https://go-review.googlesource.com/c/go/+/197938 fixed this issue in that change from entersyscallblock() to entersyscall() so we'll need to backport Minux's CL to both Go1.12 and Go1.13 as per https://groups.google.com/d/msg/golang-dev/UYDTDWHJsC8/a1_4KiKnCwAJ.

@gopherbot please backport this issue to Go1.12 and Go1.13.

**Comment by networkimprov** [comment]
@odeke-em also need to revert https://github.com/golang/go/commit/4c8037b278448af8ef38366283c82df573a2ab27

**Comment by odeke-em** [comment]
Liam, thanks for pointing that out. Minux already did that just above, but
that test as it was has a very high watermark that didn't catch the issue
until very late so we'll also need to review and update it.

**Comment by networkimprov** [comment]
I don't follow you... The commit I referenced dropped the test on the 1.13 branch. The test has been reinstated on tip.

**Comment by ianlancetaylor** [comment]
We don't need to re-enable the test on the branch.  That's not a bug fix.

**Comment by networkimprov** [comment]
Oh. I thought it hadn't shipped in 1.13.0, but I guess it did.

**Comment by networkimprov** [comment]
The test has now been re-enabled on 1.13 branch, fyi.
